opengever/maintenance/scripts/export_and_close_dossiers.py

```diff
@@ -0,0 +1,213 @@
+        unresolvable_dossiers = []
+        dossiers_in_bad_state = []
+        for brain in self.dossiers:
+            if brain.review_state in self.allowed_final_states:
+                continue
+            elif brain.review_state not in self.states_to_resolve:
+                dossiers_in_bad_state.append(brain)
+                continue
+
+            obj = brain.getObject()
+            resolver = getAdapter(obj, IDossierResolver, name="lenient")
+            try:
+                resolver.raise_on_failed_preconditions()
+            except:
+                unresolvable_dossiers.append(brain)
+
+        if unresolvable_dossiers or dossiers_in_bad_state:
+            self.log_and_write_table(
+                unresolvable_dossiers,
+                "Unresolvable dossiers",
+                "unresolvable_dossiers")
+
+            self.log_and_write_table(
+                dossiers_in_bad_state,
+                "Dossiers in bad review states",
+                "dossiers_in_bad_state")
+
+            raise PreconditionsError("Preconditions not satisfied")
+
+        logger.info("Preconditions satisfied.")
+
+    def resolve_dossiers(self):
+        message = "Closing dossiers..."
+        for brain in ProgressLogger(message, self.dossiers, logger):
+            if brain.review_state not in self.states_to_resolve:
+                continue
+
+            obj = brain.getObject()
+            resolver = getAdapter(obj, IDossierResolver, name="lenient")
+            resolver.raise_on_failed_preconditions()
+            resolver.resolve()
+            obj.reindexObject()
+
+        logger.info("Dossiers resolved.")
+
+    def export_dossiers(self):
+        message = "Exporting dossiers."
+        for brain in ProgressLogger(message, self.dossiers, logger):
+            if brain.review_state not in self.allowed_final_states:
+                raise DisallowedReviewState("Dossier in disallowed review state")
+            self._export_dossier(brain.getObject())
+
+    def _get_output_path(self, basedir, name, ext='', i=0):
+        if i > 0:
+            output_path = os.path.join(basedir, u"{}_{}{}".format(name, i, ext))
+        else:
+            output_path = os.path.join(basedir, u"{}{}".format(name, ext))
+        if os.path.exists(output_path):
+            return self._get_output_path(basedir, name, ext, i=i+1)
+        return output_path
+
+    def _export_dossier(self, dossier):
+        folder_path = self._get_output_path(self.output_directory, dossier.title)
+        os.mkdir(folder_path)
+        res = self.catalog.unrestrictedSearchResults(
+            path=dossier.absolute_url_path(),
+            object_provides=IBaseDocument.__identifier__)
+        for brain in res:
+            doc = brain.getObject()
+            filename, ext = os.path.splitext(doc.get_filename())
+            file_path = self._get_output_path(folder_path, filename, ext)
+            with open(file_path, "w") as fout:
```

Comment:
This could be done a bit more efficiently by coping the blob on the file system, instead of reading it into memory and dumping it again.
`doc.file._blob.committed()` should give you the path to the blob in blob storage, and that could then be copied using `shutil.copy2()` (which, incidentally, should also preserve last modification date).

Reply:
Thanks, I was actually looking for a better way to copy the files.